src/tactic/move_add.lean

```diff
@@ -0,0 +1,371 @@
+1. we convert a list of pairs `bool × pexpr` to a list of pairs `bool × expr`,
+2. we use the extra input `sl : list expr` to perform the unification and sorting step
+   `move_left_or_right`,
+3. we jam the third factor inside the first two.
+-/
+meta def final_sort (lp : list (bool × pexpr)) (sl : list expr) : tactic (list expr × list bool) :=
+do
+  lp_exp : list (bool × expr) ← lp.mmap $ λ x, (do e ← to_expr x.2 tt ff, return (x.1, e)),
+  (l1, l2, l3, is_unused) ← move_left_or_right lp_exp sl [],
+  return (l1 ++ l3 ++ l2, is_unused)
+
+/-- `is_given_op op e` unifies the head term of `e` with the binary operation `op`, failing
+if it cannot. -/
+meta def is_given_op (op : expr) : expr → tactic unit
+| (expr.app (expr.app F a) b) := unify op F
+| _ := failed
+
+/-- `reorder_oper op lp boos e` converts an expression `e` to a similar looking one.
+The tactic scans the expression `e` looking for subexpressions that begin with the given binary
+operation `op`.  As soon as `reorder_oper` finds one such subexpression,
+* it extracts the "`op`-summands" in the subexpression,
+* it rearranges them according to the rules determined by `lp`,
+* it recurses into each `op`-summand.
+
+The `boos` parameter is a list of booleans.  It is keeping track of which of the inputs provided
+by `lp` is actually used to perform the rearrangements.  It is useful to report unused inputs.
+In its intended application, it is set to a list of `tt`, one for each element of `lp`, indicating
+that they all begin unused.
+
+Here are two examples:
+```lean
+#eval trace $ reorder_oper ``((=)) [(ff,``(2)), (tt,``(7))] [tt] `(∀ x y : ℕ, 2 = 0)
+--  (ℕ → ℕ → 0 = 2, [ff, tt])
+-- the input `[(ff,``(2)), (tt,``(7))]` instructs Lean to move `2` to the right and `7`
+-- to the right.  Lean reports that `2` is not unused and `7` is unused as `[ff, tt]`.
+
+#eval trace $ reorder_oper ``((+)) [(ff,``(2)), (tt,``(5))] [tt, tt]
+  `(λ (e : ℕ), ∀ (x : ℕ), ∃ (y : ℕ),
+      2 + x * (y + (e + 5)) + y = x + 2 + e → 2 + x = x + 5 + (2 + y))
+/-  `2` moves to the right, `5` moves to the left.  Lean reports that `2, 5` are not unused
+    as `[ff,ff]`
+   (λ (e : ℕ), ∀ (x : ℕ), ∃ (y : ℕ),
+      x * (5 + y + e) + y + 2   = x + e + 2 → x + 2 = 5 + x + y + 2, [ff, ff]) -/
+```
+-/
+meta def reorder_oper (op : pexpr) (lp : list (bool × pexpr)) :
+  list bool → expr → tactic (expr × list bool)
+| lu F'@(expr.app F b) := (do
+    op ← to_expr op tt ff,
+    is_given_op op F',
+    (sort_list, is_unused) ← list_binary_operands op F' >>= final_sort lp,
+    sort_all ← sort_list.mmap $ reorder_oper ([lu, is_unused].transpose.map list.band),
+    let (recs, list_unused) := sort_all.unzip,
+    let summed := (recs.drop 1).foldl (λ e f, op.mk_app [e, f]) ((recs.nth 0).get_or_else `(0)),
+    return (summed, list_unused.transpose.map list.band)) <|>
+    (do [(Fn, unused_F), (bn, unused_b)] ← [F, b].mmap $ reorder_oper lu,
+    return $ (expr.app Fn bn, [unused_F, unused_b, lu].transpose.map list.band))
```

Comment:
This version with `try_core` is a little clearer than `<|>`, and also makes it possible for error messages to escape:

```suggestion
| lu F'@(expr.app F b) := do
    op ← to_expr op tt ff,
    is_op ← try_core (is_given_op op F'),
    match is_op with
    | some () := do
        (sort_list, is_unused) ← list_binary_operands op F' >>= final_sort lp,
        sort_all ← sort_list.mmap $ reorder_oper ([lu, is_unused].transpose.map list.band),
        let (recs, list_unused) := sort_all.unzip,
        recs_0 :: recs_rest ← pure recs | fail!"internal error: cannot have 0 operands",
        let summed := recs_rest.foldl (λ e f, op.mk_app [e, f]) recs_0,
        return (summed, list_unused.transpose.map list.band)
    | none := do
        [(Fn, unused_F), (bn, unused_b)] ← [F, b].mmap $ reorder_oper lu,
        return $ (expr.app Fn bn, [unused_F, unused_b, lu].transpose.map list.band)
    end
```